api/cas/auth.py

```diff
@@ -0,0 +1,303 @@
+        user.suffix = provider['user'].get('suffix')
+        user.date_last_login = timezone.now()
+        user.save()
+
+        # User must be saved in order to have a valid _id
+        user.register(username)
+        send_mail(
+            to_addr=user.username,
+            mail=WELCOME_OSF4I,
+            mimetype='html',
+            user=user
+        )
+
+    if not user.is_affiliated_with_institution(institution):
+        user.affiliated_institutions.add(institution)
+        user.save()
+
+    return user, None
+
+
+def handle_login(user):
+
+    email = user.get('email')
+    remote_authenticated = user.get('remoteAuthenticated')
+    verification_key = user.get('verificationKey')
+    password = user.get('password')
+
+    # check if credentials are provided
+    if not email or not (remote_authenticated or verification_key or password):
+        return None, MISSING_CREDENTIALS
+
+    # retrieve the user
+    user = OSFUser.objects.filter(Q(username=email) | Q(emails__icontains=email)).first()
+    if not user:
+        return None, ACCOUNT_NOT_FOUND
+
+    # verify user status
+    if verify_user_status(user) == USER_NOT_CLAIMED:
+        return None, USER_NOT_CLAIMED
+
+    # by remote authentication
+    if remote_authenticated:
+        return user, None
+
+    # by verification key
+    if verification_key:
+        if verification_key == user.verification_key:
+            return user, None
+        return None, INVALID_VERIFICATION_KEY
+
+    # by password
+    if password:
+        if user.check_password(password):
+            return user, None
+        return None, INVALID_PASSWORD
+
+
+def handle_register(user):
+
+    fullname = user.get('fullname')
+    email = user.get('email')
+    password = user.get('password')
+    if not (fullname and email and password):
+        return None, 'MISSING_CREDENTIALS'
+
+    campaign = user.get('campaign')
+    if campaign and campaign not in campaigns.get_campaigns():
+        campaign = None
+    try:
+        user = register_unconfirmed(
+            email,
+            password,
+            fullname,
+            campaign=campaign,
+        )
+    except DuplicateEmailError:
+        return None, 'ALREADY_REGISTERED'
+
+    send_confirm_email(user, email=user.username)
+    return user, None
+
+
+def get_user_with_two_factor(user):
+    return UserSettings.objects.filter(owner_id=user.pk).first()
+
+
+def verify_two_factor(user, one_time_password):
+    if not one_time_password:
+        return TWO_FACTOR_AUTHENTICATION_REQUIRED
+
+    two_factor = get_user_with_two_factor(user)
+    if two_factor and two_factor.verify_code(one_time_password):
+        return None
+    return INVALID_ONE_TIME_PASSWORD
+
+
+def verify_user_status(user):
+    """
+    Verify users' status.
+
+    Possible User Status during authentication:
+
+                        registered      confirmed       disabled        merged      usable password     claimed
```

Comment:
Here is a note on how to decide  a user's status.
- `USER_NOT_CONFIRMED` vs `USER_NOT_CLAIMED`: they only differ in whether there is a useable password
- `USER_DISABLE` and `USER_MERGED`: although several fields may vary, they can be determined only by one dedicated field: `is_disabled` and `is_merged`
- `is_claimed` is confusing and not useful